src/codeactions.ts

```diff
@@ -26,19 +26,57 @@ export class CodeActions implements vscode.CodeActionProvider {
+			.map(codeAction => codeAction as vscode.CodeAction);
 	}
 
 	private createMetadataForKey(document: vscode.TextDocument, diagnostic: vscode.Diagnostic, range: vscode.Range | vscode.Selection): vscode.CodeAction {
-		const message = this.messageList?.getMessageAt(document.offsetAt(range.start)) as Key | undefined;
+		const messageKey = this.messageList?.getMessageAt(document.offsetAt(range.start)) as Key | undefined;
 
-		const fix = new vscode.CodeAction(`Add metadata for key '${message?.value}'`, vscode.CodeActionKind.QuickFix);
+		const fix = new vscode.CodeAction(`Add metadata for key '${messageKey?.value}'`, vscode.CodeActionKind.QuickFix);
 		fix.edit = new vscode.WorkspaceEdit();
-		fix.edit.insert(document.uri, document.positionAt(message?.endOfMessage ?? 0), `,\n${' '.repeat(this.messageList?.indentation ?? 0)}"@${message?.value}" : {}`);
+		fix.edit.insert(document.uri, document.positionAt(messageKey?.endOfMessage ?? 0), `,\n${this.messageList?.getIndent()}"@${messageKey?.value}" : {}`);
 		return fix;
 	}
+
+	private createPlaceholder(document: vscode.TextDocument, diagnostic: vscode.Diagnostic, range: vscode.Range | vscode.Selection): vscode.CodeAction | undefined {
+		const placeholder = this.messageList?.getMessageAt(document.offsetAt(range.start)) as Placeholder | undefined;
+		var parent = placeholder?.parent;
+		while (!(parent instanceof MessageEntry)) {
+			parent = parent?.parent;
+		}
+		const fix = new vscode.CodeAction(`Add metadata for placeholder '${placeholder?.value}'`, vscode.CodeActionKind.QuickFix);
+		fix.edit = new vscode.WorkspaceEdit();
+
+		const parentKey = (parent as MessageEntry).key;
+		const metadatas = this.messageList?.metadataEntries.filter((entry) => entry.key.value === '@' + parentKey.value);
+		if (metadatas) {
+			const metadataForMessage = metadatas[0];
+
+			const metadata = metadataForMessage.message as Metadata;
+			if (metadata.placeholders.length > 0) {
+				const lastPlaceholderEnd = metadata.placeholders[metadata.placeholders.length - 1].objectEnd;
+				fix.edit.insert(document.uri, document.positionAt(lastPlaceholderEnd!), `,\n${this.messageList!.getIndent(2)}"${placeholder?.value}": {}`);
+			} else if (metadata.lastPlaceholderEnd) {
+				fix.edit.insert(document.uri, document.positionAt(metadata.lastPlaceholderEnd), `\n${this.messageList!.getIndent(2)}"${placeholder?.value}": {}\n${this.messageList!.getIndent(1)}`);
+			} else {
+				const insertable = `\n${this.messageList!.getIndent(1)}"placeholders": {\n${this.messageList!.getIndent(2)}"${placeholder?.value}": {}\n${this.messageList!.getIndent(1)}}\n${this.messageList!.getIndent()}`;
+				fix.edit.insert(document.uri, document.positionAt(metadata.metadataEnd), insertable);
+			}
+			return fix;
+		} else {
+			// fix.edit.insert(document.uri, document.positionAt(parentKey?.endOfMessage ?? 0), `,\n${' '.repeat(this.messageList?.this.messageList!.getIndent() ?? 0)}"@${parentKey?.value}" : {}`);
```

Comment:
Is this meant to be commented out? I assumed it would be needed by the test "Test quickfix for placeholder without metadata", but perhaps `metadatas` is empty and not undefined in that case. In which case, this probably only occurs if `messageList` is undefined - so maybe the `else` block can be removed entirely?

Reply:
Yeah, this is the case about what do when no metadata is there at all...